### DISCUSSION TITLE
How to get property of union type

### DISCUSSION THREAD

**Discussion by Jampire** [spatie/laravel-data#379] [Q&A] [answered]
Hi
Let's say I have such data class:
```php
<?php

class AlbumData extends Data
{
    public function __construct(
        public Lazy|Label $label,
        #[DataCollectionOf(SongData::class)]
        public Lazy|DataCollection $songs,
    ) {
    }

    public static function fromRequest(Request $request): self
    {
        return new self(
            label: Lazy::create(
                fn() => LabelData::from(Label::findOrNew($request->label_id))
            ),
            songs: Lazy::create(
                fn() => SongData::collection(Song::whereIn('id', $request->collect('song_ids'))->get())
            ),
        );
    }
    
    public static function fromModel(Album $album): self
    {
        return new self(
            label: $album->label,
            songs: SongData::collection($album->songs)
        );
    }
}
````

So, as you can see, if data is loaded from request, than all properties are lazy. And if data loaded from model, all properties are not lazy.

Let's say, I want to get properties ids somewhere in the code. If I do:

```php
$albumData->label->id;
$albumData->songs->toCollection()->pluck('id');
```
I will get the errors in case data was created from request: 
```
Undefined property: Spatie\LaravelData\Support\Lazy\DefaultLazy::$id
Call to undefined method Spatie\LaravelData\Support\Lazy\DefaultLazy::toCollection()
```

But if I do:
```php
$albumData->label->resolve()->id;
$albumData->songs->resolve()->toCollection()->pluck('id');
```
I will get the errors in case data was created from model: 
```
Call to undefined method AlbumData::resolve()
Call to undefined method Spatie\LaravelData\DataCollection::resolve()
```

Is there any way provided by this package to get property of union type (Lazy type plus custom or primitive type)?

**Comment by rubenvanassche** [comment]
Such thing is not possible in PHP, so the answer is sadly no

**Reply by Jampire** [reply]
@rubenvanassche , Thanks for your reply. But it's not a question of PHP. Package documentation has union types in some chapters. If it's not possible, why documentation implicitly says that it can be done?  It's already done for data transformers in `Spatie\LaravelData\Transformers\DataTransformer::resolvePropertyValue()`. So, my solution is the same:
```php
// helpers.php
if (! function_exists('resolve_data')) {
    function resolve_data(mixed $obj): mixed
    {
        return $obj instanceof Lazy ? $obj->resolve() : $obj;
    }
}
```
```php
// somewhere in the code
resolve_data($albumData->label)->id;
resolve_data($albumData->songs)->toCollection()->pluck('id');
```
But I don't like such solution. So, my question is will something be implemented by the package?

**Reply by rubenvanassche** [reply]
I'm not following what you want to see exactly? The `resolve_data` function?

**Reply by Jampire** [reply]
Let's say we have such DTO property:
```php
public Lazy|DataCollection $songs
```
How can I fetch just `songs` in my code? DTO **should not** know how it was created, right? So, I just need to write `$data->songs` and it should work regardless on how songs were created (either `DataCollection`, or `Lazy`). 

So, I see that it's almost impossible (it's done when DTO is used as response)  to do in the current implementation of the package. In this case, documentation should be changed to something like this: _Despite the fact that package can work with union type internally, you, as developer, can't manually fetch data with union type (you first need to run `instanceof` check)_.

**Comment by rubenvanassche** [comment]
> How can I fetch just songs in my code? DTO should not know how it was created, right? So, I just need to write $data->songs and it should work regardless on how songs were created (either DataCollection, or Lazy).

This is something we cannot technically implement. Only when we resolve the whole data structure when calling include but such a thing is a) a massive breaking change b) something I don't now how it exactly would work, since you can call include multiple times and we don't want to have a performance hit from resolving c) even if we can resolve the complete structure, these includes are always manually defined so it is always possible a property is still Lazy. 

So you should always manually check if something is Lazy or not.

>  In this case, documentation should be changed to something like this: Despite the fact that package can work with union type internally, you, as developer, can't manually fetch data with union type (you first need to run instanceof check).

I don't think this should be added to the docs, since you're defining a union type, you as the PHP developer should know the value can be different types and you should check them all.

**Reply by Jampire** [reply, marked as answer]
@rubenvanassche I created PR #403 with the fix of this issue. It turned out to be easier than I thought.
